### PR TITLE
Refactor how ASLR and implement rz_sys_execv using RzSubprocess

### LINKED ISSUE #4147: ASLR handling on windows is not supported in rz_sys_aslr
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | windows 10
| File format of the file you reverse (mandatory)      | PE,
| Architecture/bits of the file (mandatory)            | x64
| `rizin -v` full output, **not truncated** (mandatory)  |rizin 0.7.0 @ windows-x86-64 commit: b2112ddcdf6be4fc292a699d0f77ed3add50d682

### Expected behavior

When I rerun the program, I would like to have the same base address for any function.

### Actual behavior

The base address of roblox change sometimes.

### Steps to reproduce the behavior

I did:

```
.\OUT\bin\rizin.exe -R "aslr=no" -e "pdb.autoload=true" -e scr.color=2 -c "dc" -e "search.in=analysis.fcn" -d C:\Users\yoshi\AppData\Local\Roblox\Versions\version-88cfc23f4e7d4e4b\RobloxPlayerBeta.exe
```

### Additional Logs, screenshots, source code,  configuration dump, ...

Drag and drop zip archives containing the Additional info here, don't use external services or link.

### PR DESCRIPTION
**Pull request by wargio**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Refactors the code and moves the code into the same function.

Adds support to windows via `SetProcessMitigationPolicy` but works only when [ASLR is disable and we want to enable it](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-setprocessmitigationpolicy#remarks).

Also now on linux we try to disable ASLR programmatically using the `personality` function instead of OS wide which works also when the user is not root.

These changes also implements non-native `rz_sys_execv` by using `RzSubprocess`.

**Test plan**

Unfortunately this cannot be tested easily in the regression suite.

- Build the following C code by `gcc example.c`
```c
#include <stdio.h>

const char *test = "test";

int main() {
  printf("%p\n", (void*)test);
  return 0;
}
```

- Then create the following rz-run file `noaslr.rz`:

```bash
aslr=no
program=a.out
```

Then run `rz-run noaslr.rz` multiple times.

If the value output is the same then ASLR is correctly disabled.

**Closing issues**

Fix #4147

### CODE REVIEW

**Comment by wargio** [comment]
@gogo2464 if you want to try this on windows, it could be useful.

**Comment by wargio** [comment]
i'm currently testing this on windows.

**Comment by wargio** [comment]
So i have tested this extensively on windows and you cannot disable ASLR programmatically, only enable it.

**Comment by XVilka** [comment]
> So i have tested this extensively on windows and you cannot disable ASLR programmatically, only enable it.

This makes perfect sense from the security PoV.

File: librz/util/sys.c [hunk 1]
```diff
@@ -556,48 +610,154 @@ RZ_API bool rz_sys_chdir(RZ_NONNULL const char *s) {
 /**
```
Comment (XVilka) [inline librz/util/sys.c:610]: Add a note about Windows limitation here

File: librz/util/sys.c [hunk 2]
```diff
@@ -556,48 +610,154 @@ RZ_API bool rz_sys_chdir(RZ_NONNULL const char *s) {
...
+	}
+	if (!setProcessMitigationPolicy(ProcessASLRPolicy, &policy, sizeof(policy))) {
+		DWORD dw = GetLastError();
+		char *err = sys_windows_error_msg(dw);
+		RZ_LOG_ERROR("Cannot set ProcessASLRPolicy: %s (0x%x)\n", err, dw);
+		free(err);
+		return false;
+	}
+#elif __APPLE__
+	// Apple supports this by using the spawn api.
```
Comment (XVilka) [inline librz/util/sys.c]: Most other multiline comments are `/**/`. Lets be consistent.

File: librz/util/sys.c [hunk 3]
```diff
@@ -1501,7 +1643,45 @@ RZ_API int rz_sys_execv(const char *pathname, char *const argv[]) {
 }
 #elif !HAVE_EXECV
 RZ_API int rz_sys_execv(const char *pathname, char *const argv[]) {
```
Comment (XVilka) [inline librz/util/sys.c]: Could you please add a Doxygen to this function while you are at it?

**Review by XVilka — CHANGES_REQUESTED** [review]
These files are an absolute mess; thanks for making them a bit more readable. As for other platforms, maybe @arrowd and @devnexen could help us?

**Comment by XVilka** [comment]
It breaks one unit test on Windows:
```
 66/119 rizin:unit / run                                   FAIL             0.02s   (exit status 4294967295 or signal 4294967167 SIGinvalid)                 
```

**Comment by devnexen** [comment]
> These files are an absolute mess; thanks for making them a bit more readable. As for other platforms, maybe @arrowd and @devnexen could help us?

it builds and has the same succesful unit tests than `dev` on haiku.

**Comment by XVilka** [comment]
Static build fails subprocess unit test:
```
Summary of Failures:

 90/119 rizin:unit / subprocess                            FAIL            0.01s   exit status 1
```

**Review by XVilka — APPROVED** [review]

**Comment by gogo2464** [comment]
> So i have tested this extensively on windows and you cannot disable ASLR programmatically, only enable it.

does `-R aslr=no` works now?

**Comment by arrowd** [comment]
Is `rz_sys_aslr()` intended to change a global system-wide ASLR setting?

**Comment by gogo2464** [comment]
> Is `rz_sys_aslr()` intended to change a global system-wide ASLR setting?

only for the current process, check: https://github.com/rizinorg/rizin/blob/d86e8aebade8bdc232d09d1699ad5969e794f9c1/librz/util/sys.c#L762

**Comment by wargio** [comment]
> > So i have tested this extensively on windows and you cannot disable ASLR programmatically, only enable it.
> 
> does `-R aslr=no` works now?

on windows no and never will.

```c
		RZ_LOG_ERROR("On Windows, ASLR mitigation policies cannot be made less restrictive if they are already enabled.\n");
		RZ_LOG_ERROR("It is possible to disable ASLR by modifying the PE header and removing the following flags:\n");
		RZ_LOG_ERROR("- IMAGE_DLLCHARACTERISTICS_DYNAMIC_BASE\n");
		RZ_LOG_ERROR("- IMAGE_DLLCHARACTERISTICS_HIGH_ENTROPY_VA.\n");
		RZ_LOG_ERROR("Or as Administrator from PowerShell by running: `Set-ProcessMitigation -Name file.exe -Disable ForceRelocateImages`\n");
```

**Comment by wargio** [comment]
> Is `rz_sys_aslr()` intended to change a global system-wide ASLR setting?

it is supposed to be used for spawning processes with aslr or not.
that function will change in the future and will be only used in the debugger.

**Comment by arrowd** [comment]
Then the FreeBSD case is wrong. Changing sysctl is global and affects all processes that are started after the switch.

On FreeBSD it is possible to do `elfctl -e +noaslr file` to force an executable file to be started without ASLR. Pretty much, just like the Windows case.

**Comment by gogo2464** [comment]
> > > So i have tested this extensively on windows and you cannot disable ASLR programmatically, only enable it.
> > 
> > 
> > does `-R aslr=no` works now?
> 
> on windows no and never will.
> 
> ```c
> 		RZ_LOG_ERROR("On Windows, ASLR mitigation policies cannot be made less restrictive if they are already enabled.\n");
> 		RZ_LOG_ERROR("It is possible to disable ASLR by modifying the PE header and removing the following flags:\n");
> 		RZ_LOG_ERROR("- IMAGE_DLLCHARACTERISTICS_DYNAMIC_BASE\n");
> 		RZ_LOG_ERROR("- IMAGE_DLLCHARACTERISTICS_HIGH_ENTROPY_VA.\n");
> 		RZ_LOG_ERROR("Or as Administrator from PowerShell by running: `Set-ProcessMitigation -Name file.exe -Disable ForceRelocateImages`\n");
> ```

might we consider patching the PEB for disabling aslr?

**Comment by arrowd** [comment]
> might we consider patching the PEB for disabling aslr?

I'd say it makes sense, because it follows all other plaforms except Linux and MacOS for which it is possible to set ASLR state on the program startup.

**Comment by devnexen** [comment]
> Then the FreeBSD case is wrong. Changing sysctl is global and affects all processes that are started after the switch.
> 
> On FreeBSD it is possible to do `elfctl -e +noaslr file` to force an executable file to be started without ASLR. Pretty much, just like the Windows case.

or you can use [the procctl api with PROC_ASLR_CTL](https://man.freebsd.org/cgi/man.cgi?procctl(2)).

**Pull request merged**